### PR TITLE
offline access token, logger per microservice, no-store

### PR DESCRIPTION
**Pull request by ryanmelt**
Ready for review, but still debugging

### CODE REVIEW

File: openc3/lib/openc3/microservices/microservice.rb [hunk 1]
```diff
@@ -85,11 +86,13 @@ def initialize(name, is_plugin: false)
...
       $openc3_scope = @scope
-      Logger.scope = @scope
       @cancel_thread = false
       @metric = Metric.new(microservice: @name, scope: @scope)
+      Logger.scope = @scope
       Logger.microservice_name = @name
-      Logger.tag = @name + "__openc3.log"
+      @logger = Logger.new
+      @logger.scope = @scope
+      @logger.microservice_name = @name
```
Comment (jmthomas) [inline openc3/lib/openc3/microservices/microservice.rb:95]: What was the rationale for making this an instance variable?

File: openc3/lib/openc3/migrations/20221210174900_convert_to_multi.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
+require 'openc3/utilities/migration'
+require 'openc3/models/scope_model'
+require 'openc3/models/target_model'
+
+module OpenC3
+  class ConvertToMulti < Migration
```
Comment (jmthomas) [inline openc3/lib/openc3/migrations/20221210174900_convert_to_multi.rb:6]: This wasn't in the previous PR?

**Review by jmthomas — APPROVED** [review]
How does the offline access work in general?

File: openc3-cosmos-init/plugins/packages/openc3-cosmos-demo/targets/SYSTEM/target.txt [hunk 3]
```diff
@@ -1,12 +0,0 @@
...
-IGNORE_PARAMETER CONFIG
-IGNORE_PARAMETER OPENC3_VERSION
-IGNORE_PARAMETER USER_VERSION
-IGNORE_PARAMETER RUBY_VERSION
-
-# Ignored Items
-IGNORE_ITEM PKTID
-IGNORE_ITEM RECEIVED_COUNT
-IGNORE_ITEM RECEIVED_TIMESECONDS
-IGNORE_ITEM RECEIVED_TIMEFORMATTED
```
Comment (jmthomas) [inline openc3-cosmos-init/plugins/packages/openc3-cosmos-demo/targets/SYSTEM/target.txt:12]: Let's be sure to make these same changes in the Enterprise

File: openc3-cosmos-init/plugins/packages/openc3-tool-common/src/services/cable.js [hunk 4]
```diff
@@ -44,4 +44,7 @@ export default class Cable {
       )
     })
   }
+  recordPing() {
+    this._cable.connection.monitor.recordPing()
+  }
```
Comment (jmthomas) [inline openc3-cosmos-init/plugins/packages/openc3-tool-common/src/services/cable.js:49]: Did this have something to do with another issue? I remember a conversation about ping and the connection timeout.

File: openc3/lib/openc3/microservices/periodic_microservice.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+# This file may also be used under the terms of a commercial license
+# if purchased from OpenC3, Inc.
+
+require 'openc3/microservices/microservice'
+require 'openc3/models/offline_access_model'
+
+module OpenC3
+  class PeriodicMicroservice < Microservice
+    STARTUP_DELAY_SECONDS = 2 * 60 # Two Minutes
+    SLEEP_PERIOD_SECONDS = 24 * 60 * 60 # Run once per day
```
[the remark this answers is not in the document]
Comment (jmthomas) [reply openc3/lib/openc3/microservices/periodic_microservice.rb:25]: Does this allow a user to schedule something 30 days in the future without logging in? My general comment was "what's the point of all this". :-)

[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/microservices/periodic_microservice.rb:25]: It allows you to schedule something any time in the future.   The offline tokens expire after 30 days if they aren't used.  This makes sure they don't expire.

**Pull request merged**